### PR TITLE
[FX] Make output a non-special Node

### PR DESCRIPTION
**Pull request by jamesr66a**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45775 [FX] Track use nodes in Node
* #45227 [FX] Mutable Graph APIs
* **#45599 [FX] Make output a non-special Node**

This will normalize how function outputs are handled to match out function inputs (i.e. outputs are now represented as `output` nodes, analogous to how inputs are handled as `placeholder` nodes). This makes things more uniform in several places. `Graph.output()` now inserts an `output` node into the Graph, rather than setting a special attribute.

Differential Revision: [D24027586](https://our.internmc.facebook.com/intern/diff/D24027586)

### CODE REVIEW

File: torch/fx/graph.py [hunk 1]
```diff
@@ -80,6 +80,8 @@ def graph_copy(self, g : 'Graph'):
         """
         val_map : Dict[Node, Node] = {}
         for node in g._nodes:
+            if node.op == 'output':
+                continue
```
Comment (zdevito) [inline torch/fx/graph.py:84]: `graph_copy` is a bit unfinished: it probably should take val_map as argument, and then return the output node of the graph being copied. If that were so this would turn into a return.

File: torch/fx/graph.py [hunk 2]
```diff
@@ -157,8 +159,20 @@ def node_copy(self, node: Node, arg_transform: Callable[[Node], Argument] = lamb
         return self.create_node(node.op, node.target, args, kwargs, name)
 
     def output(self, result: Argument):
-        self.result = result
         self._mark_uses(result)
+        return self.create_node(op='output', target='output', args=(result,))
+
+    @property
```
Comment (zdevito) [inline torch/fx/graph.py:165]: I feel like this function should not exist in the version where `output` is not special. Especially as a property it suggests it is fast, but is actually linear in the size of the graph.

File: torch/fx/graph.py [hunk 3]
```diff
@@ -236,10 +251,18 @@ def python_code(self, root_module: str) -> Tuple[str, str, List[str]]:
                 assert isinstance(node.target, str)
                 body.append(f'{node.name} = {_format_target(root_module, node.target)}\n')
                 continue
+            elif node.op == 'output':
```
Comment (zdevito) [inline torch/fx/graph.py:254]: The separation of python code from the function has gotten weaker over time. We should consider just having python code generating the whole function, and that would allow us to just generate 'return out' as a string here without the special handling.

File: torch/fx/proxy.py [hunk 4]
```diff
@@ -10,6 +10,9 @@
 class TracerBase:
     graph: Graph
 
+    def output(self, result : Argument) -> Node:
```
Comment (zdevito) [inline torch/fx/proxy.py:13]: It was intentional that there only be a `create_node` function in the tracer. So this function should probably not exist.

File: torch/fx/symbolic_trace.py [hunk 5]
```diff
@@ -149,7 +149,7 @@ def module_call_wrapper(mod, *args, **kwargs):
                 return _create_proxy(self, 'call_module', module_qualified_name, args, kwargs)
         try:
             torch.nn.Module.__call__ = module_call_wrapper
-            self.graph.output(self.create_arg(fn(*args)))
+            self.output(self.create_arg(fn(*args)))
```
Comment (zdevito) [inline torch/fx/symbolic_trace.py:152]: This should call tracers create_node instead similar to how other things in symbolic trace work.

File: torch/fx/graph.py (hunk 2 above)
Comment (jamesr66a) [reply torch/fx/graph.py:165]: I agree with that, but one thing this is hiding is the fact that the output value is actually `args[0]` in the `output` node. I don't think that's nice and people will probably mess that up a lot if they have to manually extract the output value

File: test/fx/quantization.py [hunk 6]
```diff
@@ -240,7 +246,9 @@ def load_arg(a):
             if node.name in self.quants:
                 self.quants[node.name].observe(node, env)
 
-        return load_arg(self.graph.result)
+        if output_node is None:
```
Comment (zdevito) [inline test/fx/quantization.py:249]: Technically this could just go into the `if op == 'output'` statement, and cause the entire loop to terminate and return. It skips the check that there is only one return node, but we should be enforcing that anyway.

File: torch/fx/graph.py [hunk 7]
```diff
@@ -236,10 +242,22 @@ def python_code(self, root_module: str) -> Tuple[str, str, List[str]]:
...
+                if result_value:
```
Comment (zdevito) [inline torch/fx/graph.py:246]: It'd be conceptually simpler to just have output insert a `return x` into the body and delete the return formatting below. I wouldn't worry about enforcing single return semantics here, that is what the lint pass should be doing.

File: torch/quantization/fx/quantize.py [hunk 8]
```diff
@@ -485,7 +493,9 @@ def is_observed(input_arg):
                     env[node.name] = observed_graph.create_node('call_module', observer_name, (load_arg(node),), {})
                     observed_node_names_set.add(node.name)
 
-        observed_graph.output(load_arg(model.graph.result))
+        if output_node is None:
+            raise RuntimeError('Graph had no output node!')
+        observed_graph.output(load_arg(output_node.args[0]))
```
Comment (zdevito) [inline torch/quantization/fx/quantize.py:498]: Why not just put this in the loop?

File: torch/quantization/fx/quantize.py [hunk 9]
```diff
@@ -718,27 +734,38 @@ def is_quantized(node):
...
 
+        if output_node is None:
+            raise RuntimeError('Graph had no output node!')
         if is_standalone_module:
             # result are kepted quantized in the quantized standalone module
-            graph_output = map_arg(model.graph.result, load_x)
+            graph_output = map_arg(output_node.args[0], load_x)
         else:
-            graph_output = map_arg(model.graph.result, load_non_quantized)
+            graph_output = map_arg(output_node.args[0], load_non_quantized)
```
Comment (zdevito) [inline torch/quantization/fx/quantize.py:743]: Again I think this can just go into the loop, right?

File: torch/quantization/fx/quantize.py [hunk 10]
```diff
@@ -796,7 +829,9 @@ def load_arg(a):
             else:
                 # copy other nodes
                 env[node.name] = folded_graph.node_copy(node, load_arg)
-        folded_graph.output(load_arg(quantized_graph.result))
+        if output_node is None:
+            raise RuntimeError('Graph had no output node!')
+        folded_graph.output(load_arg(output_node.args[0]))
```
Comment (zdevito) [inline torch/quantization/fx/quantize.py:834]: same here

File: torch/quantization/fx/quantize.py (hunk 8 above)
Comment (jamesr66a) [reply torch/quantization/fx/quantize.py:498]: The output value is actually also used at line 502. Not sure why this seemed to pass when I ran this in FBCode...

File: torch/fx/graph.py [hunk 11]
```diff
@@ -74,13 +74,20 @@ def __init__(self):
...
+        Append all nodes from graph `g` to this graph. `val_map` should be a dictionary
+        that maps nodes in `g` to nodes in `self. `val_map` will be populated with more
+        items by this function. Returns the equivalent output value of `g` with
+        Nodes switched to refer to nodes in `self`.
         """
-        val_map : Dict[Node, Node] = {}
         for node in g._nodes:
+            if node.op == 'output':
+                rv = map_arg(node.args[0], lambda n: val_map[n])
+                assert isinstance(rv, Node)
```
Comment (zdevito) [inline torch/fx/graph.py:87]: This assert doesn't have to be true. A graph can return a tuple.

Comment (zdevito) [reply torch/fx/graph.py:87]: Return type should be `Optional[Argument]` which is just `Argument` since it already contains None.

File: torch/quantization/fx/quantize.py [hunk 12]
```diff
@@ -781,6 +791,9 @@ def load_arg(a):
         quantized_root = quantized
         quantized_graph = quantized.graph
         for node in quantized_graph.nodes:
+            if node.op == 'output':
```
Comment (zdevito) [inline torch/quantization/fx/quantize.py:794]: This is probably not needed because there is a `node_copy` in the else branch which will do the same thing.

**Review by zdevito — APPROVED** [review]
Cool -- it looks like making output a node simplifies things compared to making parameters not nodes. So I think we should go with this approach. It should also simplified the mutation logic since you can always insert after a param or insert before a return.
